File: ios/src/main/java/y2k/joyreactor/NetworkIndicator.java
package y2k.joyreactor;

import org.robovm.apple.uikit.UIApplication;
import rx.Observable;
import rx.Subscriber;

import java.util.concurrent.TimeUnit;

/**
 * Created by y2k on 10/13/15.
 */
public class NetworkIndicator {

    private Subscriber<? super Object> subscriber;

    public NetworkIndicator() {
        Observable
                .create(subscriber -> this.subscriber = subscriber)
                .buffer(500, TimeUnit.MILLISECONDS)
                .filter(s -> s.size() > 0)
                .map(s -> s.get(s.size() - 1))
                .observeOn(ForegroundScheduler.getInstance())
                .subscribe(s -> update((Boolean) s));
    }

    public void setEnabled(boolean isEnabled) {
        subscriber.onNext(isEnabled);
    }

    private void update(boolean isEnabled) {
        UIApplication.getSharedApplication().setNetworkActivityIndicatorVisible(isEnabled);
    }
}

File: ios/src/main/kotlin/y2k/joyreactor/common/NetworkIndicator.kt
package y2k.joyreactor.common

import org.robovm.apple.uikit.UIApplication
import rx.Observable
import rx.Subscriber
import java.util.concurrent.TimeUnit

/**
 * Created by y2k on 10/13/15.
 */
class NetworkIndicator {

    private lateinit var subscriber: Subscriber<in Boolean>

    init {
        Observable
            .create<Boolean> { subscriber -> this.subscriber = subscriber }
            .buffer(500, TimeUnit.MILLISECONDS)
            .filter { it.size > 0 }
            .map { it[it.size - 1] }
            .observeOn(ForegroundScheduler.instance)
            .subscribe { s -> update(s as Boolean) }
    }

    fun setEnabled(isEnabled: Boolean) {
        subscriber.onNext(isEnabled)
    }

    private fun update(isEnabled: Boolean) {
        UIApplication.getSharedApplication().isNetworkActivityIndicatorVisible = isEnabled
    }
}